Create new API version 2024-11-01 adding `provider` field to GitRepository

dipti-pai opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

diptipai89 committed:
Copy files from preview/2024-04-01-preview

Copied the files in a separate commit.
This allows reviewers to easily diff subsequent changes against the previous spec.

diptipai89 committed:
Update version to stable/2024-11-01

Updated the API version from preview/2024-04-01-preview to stable/2024-11-01.

diptipai89 committed:
Added tag for 2024-11-01 in readme file

diptipai89 committed:
Add provider field to GitRepositoryDefinition and GitRepositoryPatchDefinition

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

raosuhas commented:
Could you please check the swagger lintdiff errors. It looks like you do not have an operations api implementation for this api version ?

diptipai89 committed:
add missing apis

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

diptipai89 committed:
Fix prettier validations

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

dipti-pai commented:
@raosuhas I don't have permissions to remove the [ARMChangesRequested](https://github.com/Azure/azure-rest-api-specs/labels/ARMChangesRequested) label. Would you know how I can get the required permissions to continue the review ?

dipti-pai commented:
/azp run

dipti-pai commented:
/azp run

diptipai89 committed:
Add missing sdk readme files

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

dipti-pai commented:
/azp run

ramoka178 commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2024-11-01/examples/CreateFluxConfigurationWithOCIRepository.json:
```diff
@@ -0,0 +1,304 @@
+    "clusterRp": "Microsoft.Kubernetes",
+    "clusterResourceName": "connectedClusters",
+    "fluxConfigurationName": "srs-fluxconfig",
+    "api-version": "2024-04-01",
```
>2024-04-01 [](http://example.com/codeflow?start=20&length=10)

fix this

ramoka178 commented:
Please fix Swagger Avocado failures too

diptipai89 committed:
fix version in example directory

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

diptipai89 committed:
Fix swagger avocado error

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

dipti-pai commented:
/azp run

ramoka178 commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2024-11-01/fluxconfiguration.json:
```diff
@@ -0,0 +1,1849 @@
+          "type": "string",
+          "x-nullable": true
+        },
+        "provider": {
```
>provider [](http://example.com/codeflow?start=9&length=8)

are there any known set of providers for this to be enum ? Also update examples to reflect this new property

ramoka178 replied:
Adding new enums should be ok. You need not have a new API version for it.

ramoka178 replied:
Adding new enums should be ok. You need not have a new API version for it.

diptipai89 committed:
Convert provider to enum and add an example

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

ramoka178 commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2024-11-01/fluxconfiguration.json:
```diff
@@ -0,0 +1,1860 @@
+          "x-nullable": true,
+          "enum": [
+            "azure",
+            "generic"
```
Please make them 'Azure' and 'Generic' and add descriptions for each enum on what they mean. 

Descriptions are used to generate comments/docs and so end customer will have context on what this means.

diptipai89 committed:
Add comments for provider enums

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

ramoka178 commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2024-11-01/fluxconfiguration.json:
```diff
@@ -0,0 +1,1860 @@
+          ],
+          "x-ms-enum": {
+            "name": "ProviderType",
+            "modelAsString": true
```
Sorry, I meant to add  inside "x-ms-enum" as 

"values": [
          {
            "name": "Azure",
            "value": "Azure",
            "description": "<<<Its description>>>."
          } ,
           {
            "name": "Generic",
            "value": "Generic",
            "description": "<<<Its description>>>."
          } ]
          
You can refer other swaggers for reference

ramoka178 commented:
Please fix Swagger ModelValidation errors too

diptipai89 committed:
fix example

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

diptipai89 committed:
Add enum values with description

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

dipti-pai commented:
@ramoka178 As per the Next Steps diagram, SDK reviewers are supposed to review the changes next. It's been few days and there  have been no updates. Would you know how I can bring this PR to SDK reviewers attention ? Thanks so much!

kazrael2119 commented:
add @JeffreyRichter as reviewer as the newer default tap has multi api versions

JeffreyRichter commented:
Thanks for adding me in.
Yes, Azure's policy and processes require that all operations in a service version uniformly. So the 2024-11 package cannot contain operations with a version of 2023-05-01. ALL operations MUST be 2024-11-01.
We do have an internal document that explains why this is necessary: https://microsoft-my.sharepoint.com/:w:/p/jeffreyr/EVQ3n8VAjgxGknd19qeY_LsBf48ooc71gTiY22U1NNKnPA?e=KXhvnc

dipti-pai commented:
> Azure's policy and processes require that all operations in a service version uniformly. So the 2024-11 package cannot contain operations with a version of 2023-05-01. ALL operations MUST be 2024-11-01. 

@JeffreyRichter Is this requirement only for stable versions and not preview ?

JeffreyRichter commented:
The document explains this.
This requirement is for everything: private preview, public preview, and stable. The requirement exists so that customers know what docs version to look at, what SDK package version to get, to know what service versions are deployed in which datacenters & non-public clouds, to know what versions are boroken if a new breaking-change version is introduced, and to know which service versions are retired when they have to be retired. In other words, ALL of our internal processes and customer-facing messaging force this requirement.

This requirement has always been true, but we haven't always actively enforced it. We are now actively enforcing it due to various customer issues that have occurred and we are improving our repo tooling to detect these kinds of issues and forbid merging PRs that violate them. We don't have this tooling just yet which is why I'm asked to manually intervene when someone detects a PR in violation (like this PR).

dipti-pai commented:
Thanks @JeffreyRichter for the explanation. Our previous version (2024-04-01-preview) has diverged and has multi APIs, could you suggest (or share a doc) how I should make the new stable API 2024-11-01 using the scripts in ./eng/scripts given that fluxConfiguration.json APIs are based on 2024-04-01-preview and the other APIs (extensions, operations, kubernetesConfiguration) are based on 2023-05-01 ?

JeffreyRichter commented:
I don't know anything about the scripts in ./eng/scripts.
The teams that I've worked with on this have had to do some internal coordination (I don't know the details). There is an alternate approach of splitting a monolithic service into a suite of smaller service where each service versions uniformly but independently of each other. 
Really, what a team does should be based on the customer experience - does your team want to represent its features as a single monolithic service that versions where each version has its own docs, SDK package, etc. OR does the team want a suite of services where each has its own separate docs, SDK package, etc.? Once the decision is made, it must be maintained in perpetuity as changing this is a breaking change. The Azure Breaking Board is allowing teams to make a 1-time break to go from monolith to suite so that teams can get on a long-term sustainable path.

But again, your team should optimize for the best customer experience (not the best MS engineering experience) and then put in place a long-term plan to maintain this experience.

diptipai89 committed:
Copy files from stable/2023-05-01

Copied the files in a separate commit.
This allows reviewers to easily diff subsequent changes against the previous spec.

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

diptipai89 committed:
Update version to stable/2024-11-01
Updated the API version from stable/2023-05-01 to stable/2024-11-01.

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

diptipai89 committed:
Added tag for 2024-11-01 in readme file

Signed-off-by: Dipti Pai <diptipai@microsoft.com>

dipti-pai commented:
@kazrael2119, I removed the multi-APIs, all APIs are now in version 2024-11-01. Could you take a look? Thanks.

kazrael2119 committed:
Create sdk-suppressions.yaml

tadelesh commented:
@JeffreyRichter in this pr, there is a ref change that will cause sdk breaking change: `KustomizationPatchDefinition.PostBuild` has been changed from `PostBuildDefinition` to `PostBuildPatchDefinition`. it is only labeled with warning in swagger breaking change tool: https://github.com/Azure/azure-rest-api-specs/pull/30956/checks?check_run_id=33022160718. idw whether it should go through the breaking change review?

JeffreyRichter commented:
Was changing PostBuildDefinition to PostBuildPatchDefinition intentional?
If a mistake, then it should be put back. If intentional, then it should be justified by the service team.

dipti-pai commented:
> Was changing PostBuildDefinition to PostBuildPatchDefinition intentional? If a mistake, then it should be put back. If intentional, then it should be justified by the service team.

@tadelesh @JeffreyRichter This change is not new to this PR and is present in the previous version (2024-04-01-preview) that this version is based on. Link [here](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2024-04-01-preview/fluxconfiguration.json#L1428).

JeffreyRichter commented:
I added the BreakingChange-Approved-Previously label so this PR is good as is.

tadelesh committed:
Add suppressions for azure-sdk-for-go breaking changes

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Microsoft.KubernetesConfiguration](https://apiview.dev/Assemblies/Review/ffe759f6cde74eef843b248f0d153d07?revisionId=4131cc811d6d492bb9df2b81d3182126)
